[cli] Added cli extension helpers for building extensions

chrfalch opened the pull request:
# Why

When building CLI extensions we need some utility functions for setting up and implementing the extensions. A CLI extension is installed as part of an Expo package and should be available in debug mode and in development builds created on EAS.

# How

This PR adds a few utility functions for creating CLI extensions:

- Setup function for creating a typed cli extension
- Imperative/asynchronous function for sending a message to one or more running apps and waiting for the result value including a timeout from within the extension
- Imperative function for setting up a message listener in a running Expo app to receive messages from a cli extension

Setting up an extension:

An extension is a simple node-based command line application that is distributed as an npm package and installed as part of an expo package, configured in `expo-module.config.json`.

Creating an extension using one of the utility functions:

```ts
cliExtension<{ list: Record<string, never>; test: Record<string, never> }>(
  async (command, _args, metroServrrOrigin) => {
    const apps = await queryAllInspectorAppsAsync(metroServerOrigin);
    if (apps.length === 0) {
      throw new Error(
        'No apps connected to the dev server. Please connect an app to use this command.'
      );
    }

    if (command === 'list') {
      try {
        return [
          { type: 'text', text: await sendMessageAsync("get_registered_tasks", "expo-background-task-plugin", apps) },
        ];
      } catch (error: any) {
        throw new Error('An error occured connecting to the app:' + error.toString());
      }
    } else if (command === 'test') {
      // Trigger background tasks
      try {
        return [{ type: 'text', text: await sendMessageAsync("trigger_tasks", "expo-background-task-plugin", apps) }];
      } catch (error: any) {
        throw new Error('An error occured connecting to the app:' + error.toString());
      }
    } else {
      return Promise.reject(
        new Error("Unknown command. Use 'list' to see available tasks or 'trigger' to run a task.")
      );
    }
  }
);
```

The implementation on the app side is also trivial given the helper functions:

```ts
startDevToolsPluginListenerAsync("expo-background-task-plugin")
  .then(({ addMessageListener }) => {
    // Handle the trigger background tasks request
    addMessageListener("trigger_tasks", async ({ sendResponseAsync }) => {
      const tasks = await TaskManager.getRegisteredTasksAsync();
      if (tasks.length === 0) {
        await sendResponseAsync('No background tasks registered to trigger.');
        return;
      }
      // Trigger the background tasks
      await triggerTaskWorkerForTestingAsync();
      await sendResponseAsync(`${tasks.length} tasks triggered successfully.`);
    });

    // Handle the get registered tasks request
    addMessageListener("get_registered_tasks", async ({ sendResponseAsync }) => {
      const tasks = await TaskManager.getRegisteredTasksAsync();
      const message =
        tasks.length === 0
          ? 'No background tasks registered.'
          : `${tasks.length} task(s): ${tasks.map((task) => `"${task.taskName}"`).join(', ')}.`;
      await sendResponseAsync(message);
    });
  })
  .catch((error) => {
    console.error('Failed to start app listeners for expo-backgroundtask-devtools-plugin:', error);
  });
```

# Test Plan

Run in BareExpo with the dummy expo-background-task cli extension that is part of this stack

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[cli] Added cli extension helpers for building extensions

When building extensions we need a few tools to simplify repetivite tasks. These includes:

- Utility function for creating/defining a cli extension
- Imperative function for listening for messages in the consuming library (used in the expo app)
- tools for sending messages to a running app

chrfalch commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/38093?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#38093** <a href="https://app.graphite.dev/github/pr/expo/expo/38093?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/expo/expo/38093?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#38091** <a href="https://app.graphite.dev/github/pr/expo/expo/38091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#38092](https://github.com/expo/expo/pull/38092) <a href="https://app.graphite.dev/github/pr/expo/expo/38092?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

chrfalch committed:
[cli] added build for new utility functions

chrfalch committed:
[cli] added changelog entry

Kudo commented on packages/expo/src/devtools/CliExtensionMessages.ts:
```diff
@@ -0,0 +1,17 @@
+  ) => void;
+  sendMessageAsync: (eventName: string, message: string) => Promise<void>;
+}> => {
+  throw new Error(
```
the error may break module from web. maybe turn it to no-op function

chrfalch commented on packages/expo/src/devtools/CliExtensionSendMessageAsync.ts:
```diff
@@ -0,0 +1,153 @@
+  }
+  // Create connection
+  const url = new URL(apps[0].webSocketDebuggerUrl);
+  const address = `ws://${url.host}/expo-dev-plugins/broadcast`;
```
startDevToolsPluginListenerAsync is called/started on the device/app, while sendMessageAsync is called from the cli-extension.

chrfalch replied:
**sendMessageAsync**: called by the node cmd-line tool to send a message from the tool -> apps.
**startDevToolsPluginListener**: called by the consuming native app to listen to messages send by the node cmd-line tool

None of these are inside the cli - the node cmd-line tool is ran using a separate node instance and have no knowledge or connection to the cli's websocket.

We can chat through this one online if you need more info.

chrfalch commented on packages/expo/src/devtools/CliExtensionMessages.native.ts:
```diff
@@ -0,0 +1,96 @@
+import * as Application from 'expo-application';
+import * as DeviceInfo from 'expo-device';
```
Good point. What I need these for are for sending `Application.applicationId` and device info back when receiving a message. This is done so that the extension running can receive messages from multiple apps so that it can map it back to the connected apps in the cli.

Kudo replied:
i think we should resolve the problem through the underlying devtools plugin websocket transport. it's just that i didn' get a chance to resolve it.
wdyt to support single app in the meantime? (we also support single app for devtools plugins)

chrfalch committed:
codereview:

- Added tests for the arguments parser
- Changed to test for `process.env.NODE_ENV !== 'production'` instead of dev launcher installed
- Changed throwing an exception on non-native (android/ios) platforms to showing a warning

Kudo commented on packages/expo/src/devtools/CliExtensionMessages.native.ts:
```diff
@@ -0,0 +1,96 @@
+    clientRef = client;
+
+    // Ensure the module is properly cleaned up
+    // @ts-ignore
```
that's reasonable. then to document a little bit where the `module.hot` come from and also add a `close()` function to allow people terminate the websocket connection

chrfalch replied:
I can add some comments describing this, good!
I don't need a close() - since the listener should be open until the app is closed or reloaded.

chrfalch committed:
codereview: Refactored apps argument

Apps are now longer stringified and passed to tools. Only metroServerOrigin is passed which can then be used to query metro for connected apps.

chrfalch committed:
codereview: updated devtools to export queryAllInspectorAppsAsync

Updated types
Updated export
